### ISSUE TITLE
worker_quit_after_build flag quits workers after each action instead of entire build invocation 

### ISSUE THREAD

**Issue by nkoroste** [bazelbuild/bazel#12466]
### Description of the problem / feature request:

Originally, I thought `--worker_quit_after_build` will quit all workers at the end of the entire Bazel build invocation. Which will add ~15s of total overhead but it seems like it will quit all workers after every actions which will add this ~15s overhead per action.

### Feature requests: what underlying problem are you trying to solve with this feature?

I'm trying to use `--worker_quit_after_build` flag to reduce the number of resources Bazel is consuming to reduce the chance of OOMs. Since it looks like some workers are leaky and after multiple large builds the machine resources are depleted and everything freezes. 

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. Do a clean no remote/disk cache build with and without `--worker_quit_after_build` 
2. Compare the trace/profile file produced by each build.
3. You should notice that when the flag is used there is more overhead per action. 

### What operating system are you running Bazel on?

Linux, MacOS

### What's the output of `bazel info release`?

Last tested on Bazel v3.4.1

**Comment by larsrc-google** [comment]
Here's my command for attempting to reproduce this problem:

```
bazel --bazel_rc=/dev/null build //src/main/java/com/google/devtools/build/lib/worker:work_request_handlers --profile=/tmp/worker_profile --strategy=Javac=worker --worker_quit_after_build --worker_verbose
```

Output from the first run:

```
$ bazel --bazelrc=/dev/null build //src/main/java/com/google/devtools/build/lib/worker:work_request_handlers --profile=/tmp/worker_profile --strategy=Javac=worker --worker_quit_after_build --worker_verbose
DEBUG: /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:68:14: 
Current running Bazel is ahead of bazel-toolchains repo. Please update your pin to bazel-toolchains repo in your WORKSPACE file.
DEBUG: /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1804_java11 not using checked in configs as detect_java_home was set to True 
DEBUG: /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:68:14: 
Current running Bazel is ahead of bazel-toolchains repo. Please update your pin to bazel-toolchains repo in your WORKSPACE file.
DEBUG: /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1604_java8 not using checked in configs as detect_java_home was set to True 
INFO: Analyzed target //src/main/java/com/google/devtools/build/lib/worker:work_request_handlers (30 packages loaded, 1006 targets configured).
INFO: Found 1 target...
INFO: Created new non-sandboxed Javac worker (id 6), logging to /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/bazel-workers/worker-6-Javac.log
INFO: Created new non-sandboxed Javac worker (id 7), logging to /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/bazel-workers/worker-7-Javac.log
INFO: Created new non-sandboxed Javac worker (id 8), logging to /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/bazel-workers/worker-8-Javac.log
INFO: Created new non-sandboxed Javac worker (id 9), logging to /usr/local/google/home/larsrc/.cache/bazel/_bazel_larsrc/0a7ca9b3538362b7584391f865ed2c53/bazel-workers/worker-9-Javac.log
Target //src/main/java/com/google/devtools/build/lib/worker:work_request_handlers up-to-date:
  bazel-bin/src/main/java/com/google/devtools/build/lib/worker/libwork_request_handlers.jar
INFO: Build completed, shutting down worker pool...
INFO: Destroying Javac worker (id 7)
INFO: Destroying Javac worker (id 8)
INFO: Destroying Javac worker (id 9)
INFO: Destroying Javac worker (id 6)
INFO: Elapsed time: 75.579s, Critical Path: 46.03s
INFO: 222 processes: 16 internal, 202 linux-sandbox, 4 worker.
INFO: Build completed successfully, 222 total actions
```

As you can see, the workers are only destroyed at the end. If you run with --worker_verbose, do you see more creation/destruction of workers? What worker are you using?

Execution times:
--worker_quit_after_build: INFO: Elapsed time: 75.579s, Critical Path: 46.03s
--noworker_quit_after_build: INFO: Elapsed time: 66.157s, Critical Path: 40.90s
--worker_quit_after_build (workers warmed up): INFO: Elapsed time: 58.362s, Critical Path: 33.51s
--worker_quit_after_build: INFO: Elapsed time: 62.291s, Critical Path: 39.92s
--worker_quit_after_build: INFO: Elapsed time: 62.630s, Critical Path: 34.05s
--noworker_quit_after_build: INFO: Elapsed time: 58.972s, Critical Path: 28.48s
--noworker_quit_after_build: INFO: Elapsed time: 72.453s, Critical Path: 37.37s
--noworker_quit_after_build: INFO: Elapsed time: 59.641s, Critical Path: 40.29s

This is on an otherwise idle cloud linux machine. As you can see, there's some variation, but overall there's a couple of seconds penalty for workers that are not warmed up, which is expected. 

If you compare with and without --worker_quit_after_build, but in between do not just `bazel clean`, but also `bazel shutdown`, do you see the same difference?

**Comment by larsrc-google** [comment]
Downgrading to P2 for lack of response.

**Comment by nkoroste** [comment]
Sorry for the delayed response, here is the worker strategy we're using:
```
# Enable D8 instead of DexBuilder and DexMerger
build:d8 --define=android_dexmerger_tool=d8_dexmerger
build:d8 --define=android_incremental_dexing_tool=d8_dexbuilder
build:d8 --nouse_workers_with_dexbuilder # https://github.com/bazelbuild/bazel/issues/10241
build:d8 --incremental_dexing # to overwrite our default

# Enabling workers can result in better build performance by saving on JVM
# startup costs from invoking each of these tools, but at the cost of increased
# memory usage on the system by persisting them.
# To enable workers for these actions, apply these flags
# with --config=android_workers on the command line:
build:android_workers --strategy=Javac=worker
build:android_workers --strategy=Desugar=worker
build:android_workers --strategy=KotlinCompile=worker
build:android_workers --persistent_android_resource_processor
# For some reason AARGenerator fails when in worker mode.
build:android_workers --strategy=AARGenerator=sandboxed
# Reduce the number of workers to max 2 to avoid OOM (default is 4)
build:android_workers --worker_max_instances=2
# Bump kotlin compile workers
build:android_workers --worker_max_instances=KotlinCompile=8
# build:android_workers --strategy=DexBuilder=worker # Disabled due to https://github.com/bazelbuild/bazel/issues/10241
```

You're right, I do see similar results with `--shutdown` but it's significantly more than a couple of seconds:
Results after making a change in a medium-size build target:
Clean (no extra flags) -- `Elapsed time: 888.799s (14.81m), Critical Path: 885.08s (14.75m)`
with `shutdown` flag -- `Elapsed time: 1048.905s (17.48m), Critical Path: 1025.63s (17.09m)`
with `worker_quit_after_build` flag -- `Elapsed time: 999.393s (16.66m), Critical Path: 995.54s (16.59m)`

Building APK (no cache)
clean (no extra flags) -- `Elapsed time: 2361.239s (39.35m), Critical Path: 2103.39s (35.06m)`
with `shutdown` flag -- `Elapsed time: 2337.382s (38.96m), Critical Path: 2085.79s (34.76m)`
with `worker_quit_after_build` flag -- `Elapsed time: 2484.000s (41.40m), Critical Path: 2099.95s (35.00m)`

Did you have a chance to look at the trace file generated for each of these use cases? You should be able to see that there is significant overhead per action when `worker_quit_after_build` is used. Though I guess it will be the same if `shutdown` is used? -- I haven't verified this last bit.

**Comment by larsrc-google** [comment]
Are you still seeing this issue? 

Also, I've fixed a GC leak on Mac in the meanwhile, which might help on the overhead you mention at the start?

**Comment by nkoroste** [comment]
Can you link to the GC leak fix change? 

My findings above (https://github.com/bazelbuild/bazel/issues/12466#issuecomment-740915617) were made on a Linux machine so if the fix impacts only Mac there might still be an issue.

**Comment by larsrc-google** [comment]
The fixes are commit 6bf44ba (don't GC all the time) and commit 94ac839 (GC only occasionally). But yeah, I don't know of any impact of this on Linux.

Are you still seeing these extra worker deaths? Do they show when passing `--worker_verbose`?

If you can patch in commit 786b418, you will get details on the workers in the profiles.
